File: app/src/main/java/com/example/adam/myapplication/ui/doctor/dialogs/ChooseDoctorAdapter.java
package com.example.adam.myapplication.ui.doctor.dialogs;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.adam.myapplication.R;
import com.example.adam.myapplication.data.objects.Doctor;
import com.example.adam.myapplication.ui.OnItemClickListener;

import java.util.ArrayList;
import java.util.List;

public class ChooseDoctorAdapter extends RecyclerView.Adapter<ChooseDoctorAdapter.DoctorViewHolder> {

    private List<Doctor> list = new ArrayList<>();
    private OnItemClickListener<Doctor> listener;

    ChooseDoctorAdapter(OnItemClickListener<Doctor> listener) {
        this.listener = listener;
    }

    void loadDoctors(List<Doctor> list) {
        this.list = list;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public DoctorViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int i) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.item_choose_doctor, parent, false);

        return new DoctorViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull DoctorViewHolder holder, int position) {
        Doctor doctor = list.get(position);

        holder.setSpecialization(doctor.getSpecialization());
        holder.setName(doctor.getName());
        holder.setListener(listener);
    }


    @Override
    public int getItemCount() {
        return list.size();
    }

    class DoctorViewHolder extends RecyclerView.ViewHolder {

        private TextView specialization;
        private TextView name;

        DoctorViewHolder(@NonNull View itemView) {
            super(itemView);

            specialization = itemView.findViewById(R.id.specialization_text);
            name = itemView.findViewById(R.id.name_text);
        }

        void setListener(final OnItemClickListener<Doctor> listener) {
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listener.onItemClick(itemView, list.get(getAdapterPosition()));
                }
            });
        }

        void setSpecialization(String specialization) {
            this.specialization.setText(specialization);
        }

        public void setName(String name) {
            this.name.setText(name);
        }
    }
}

File: app/src/main/java/com/example/adam/myapplication/ui/lists/ChooseDoctorAdapter.kt
package com.example.adam.myapplication.ui.lists

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.example.adam.myapplication.R
import com.example.adam.myapplication.data.objects.Doctor

import java.util.ArrayList

class ChooseDoctorAdapter internal constructor(private val listener: OnItemClickListener<Doctor>) : RecyclerView.Adapter<ChooseDoctorAdapter.DoctorViewHolder>() {

    private var list: List<Doctor> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, i: Int): DoctorViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(R.layout.item_choose_doctor, parent, false)

        return DoctorViewHolder(view)
    }

    override fun onBindViewHolder(holder: DoctorViewHolder, position: Int) {
        val doctor = list[position]

        holder.setSpecialization(doctor.specialization ?: "")
        holder.setName(doctor.name ?: "")
        holder.setListener(listener)
    }

    override fun getItemCount(): Int {
        return list.size
    }

    fun loadDoctors(list: List<Doctor>) {
        this.list = list
        notifyDataSetChanged()
    }

    inner class DoctorViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private val specialization: TextView = itemView.findViewById(R.id.specialization_text)
        private val name: TextView = itemView.findViewById(R.id.name_text)

        fun setListener(listener: OnItemClickListener<Doctor>) {
            itemView.setOnClickListener { listener.onItemClick(itemView, list[adapterPosition]) }
        }

        fun setSpecialization(specialization: String) {
            this.specialization.text = specialization
        }

        fun setName(name: String) {
            this.name.text = name
        }
    }
}
